File: unity/logbrew-unity/Samples~/RealUserSmoke/RealUserSmoke.cs
#nullable enable

using System;
using System.Collections.Generic;
using LogBrew.Unity;

public static class RealUserSmoke
{
    public static void Main()
    {
        var client = LogBrewUnity.CreateClient("LOGBREW_API_KEY", "logbrew-unity");
        ReadmeExample.EnqueueCanonicalEvents(client);
        Console.WriteLine(client.PreviewJson());
        var response = client.Flush(new RecordingTransport(new object[] { TransportException.Network("temporary outage"), 202 }));

        var helperClient = LogBrewUnity.CreateClient("LOGBREW_API_KEY", "logbrew-unity-helper");
        var context = UnityContext.Create()
            .WithPlatform("ios")
            .WithSceneName("MainMenu")
            .WithGameObjectName("Player")
            .WithSessionId("session_001")
            .WithFrame(42);
        LogBrewUnity.CaptureSceneLoaded(helperClient, "evt_scene_loaded_001", "2026-06-02T10:00:06Z", "MainMenu", 1, context);
        LogBrewUnity.CaptureLogMessage(helperClient, "evt_unity_log_001", "2026-06-02T10:00:07Z", "button clicked", "Log", context);
        LogBrewUnity.CaptureException(helperClient, "evt_unity_exception_001", "2026-06-02T10:00:08Z", "NullReferenceException", "stack trace", context);
        var helperPreview = helperClient.PreviewJson();
        if (!helperPreview.Contains("\"sceneName\": \"MainMenu\"") || !helperPreview.Contains("\"unityLogType\": \"Log\""))
        {
            throw new InvalidOperationException("unity helper metadata missing");
        }

        var richContext = TelemetryContext.Create()
            .WithResource(TelemetryResource.Create()
                .WithDeployment("production", "2.3.0")
                .WithFramework("unity", "6000.1")
                .WithApplication("Checkout Game", "2.3.0", "204")
                .Build())
            .WithSession("session_001")
            .WithSubject("opaque_player_001", "user")
            .WithTag("journey", "checkout")
            .Build();
        var richClient = LogBrewUnity.CreateClient(
            "LOGBREW_API_KEY",
            "checkout-game",
            context: richContext,
            includeAutomaticContext: false);
        richClient.AddBreadcrumb(
            IssueBreadcrumb.Create("2026-06-02T10:00:07Z", "checkout.request")
                .WithType("http")
                .WithLevel("warning")
                .WithMessage("retry started")
                .WithData(new Dictionary<string, object?> { ["attempt"] = 2 }));
        LogBrewUnity.CaptureException(
            richClient,
            "evt_rich_issue_001",
            "2026-06-02T10:00:08Z",
            "NullReferenceException",
            "Checkout.Submit () (at /workspace/game/Assets/Scripts/Checkout.cs:42)",
            context);
        richClient.Metric(
            "evt_metric_001",
            "2026-06-02T10:00:09Z",
            MetricAttributes.Create("frame.duration", "histogram", 16.6, "ms", "delta")
                .WithMetadata(new Dictionary<string, object?> { ["scene"] = "Checkout" }));
        richClient.Span(
            "evt_span_evidence_001",
            "2026-06-02T10:00:10Z",
            SpanAttributes.Create(
                    "checkout.submit",
                    "4bf92f3577b34da6a3ce929d0e0e4736",
                    "00f067aa0ba902b7",
                    "error")
                .WithEvent(SpanEventSummary.Create("retry").WithMetadata(new Dictionary<string, object?> { ["attempt"] = 2 }))
                .WithLink(SpanLinkSummary.Create("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", "bbbbbbbbbbbbbbbb").WithSampled(false)));
        var richPreview = richClient.PreviewJson();
        foreach (var expected in new[]
        {
            "\"schemaVersion\": 1",
            "\"subject\"",
            "\"breadcrumbs\"",
            "\"filename\": \"Checkout.cs\"",
            "\"type\": \"metric\"",
            "\"events\"",
            "\"links\""
        })
        {
            if (!richPreview.Contains(expected, StringComparison.Ordinal))
            {
                throw new InvalidOperationException("rich Unity telemetry missing " + expected);
            }
        }

        if (richPreview.Contains("/workspace/game", StringComparison.Ordinal))
        {
            throw new InvalidOperationException("rich Unity telemetry leaked an absolute source path");
        }

        var httpClient = LogBrewUnity.CreateClient("LOGBREW_API_KEY", "logbrew-unity-http", maxRetries: 1);
        httpClient.Log(
            "evt_unity_http_transport",
            "2026-06-02T10:00:09Z",
            LogAttributes.Create("unity http transport sent", "info").WithLogger("unity-http"));
        var capturedAuthorization = string.Empty;
        var httpResponse = httpClient.Flush(new HttpTransport(
            new Uri("https://example.logbrew.test/v1/events"),
            new Dictionary<string, string> { ["x-logbrew-source"] = "unity-smoke" },
            TimeSpan.FromSeconds(10),
            requester: request =>
            {
                capturedAuthorization = request.Headers["authorization"];
                if (request.Headers["content-type"] != "application/json"
                    || request.Headers["x-logbrew-source"] != "unity-smoke"
                    || !request.Body.Contains("evt_unity_http_transport", StringComparison.Ordinal))
                {
                    throw new InvalidOperationException("unexpected HTTP transport request");
                }

                return string.IsNullOrEmpty(capturedAuthorization) ? 500 : 202;
            }));
        if (capturedAuthorization != "Bearer LOGBREW_API_KEY")
        {
            throw new InvalidOperationException("unexpected HTTP transport authorization");
        }

        Console.Error.WriteLine("{\"ok\":true,\"status\":" + response.StatusCode + ",\"retryAttempts\":" + response.Attempts + ",\"unityHelperEvents\":3,\"richContextEvents\":3,\"metricEvents\":1,\"httpAttempts\":" + httpResponse.Attempts + "}");
    }
}

File: kotlin/logbrew-kotlin/examples/real_user_smoke/RealUserSmoke.kt
import co.logbrew.sdk.AndroidContext
import co.logbrew.sdk.AndroidLogPriority
import co.logbrew.sdk.HttpTransport
import co.logbrew.sdk.HttpTransportRequester
import co.logbrew.sdk.LogAttributes
import co.logbrew.sdk.LogBrewAndroid
import co.logbrew.sdk.LogBrewClient
import co.logbrew.sdk.MetricAttributes
import co.logbrew.sdk.RecordingTransport
import co.logbrew.sdk.TransportException

fun main() {
    val client =
        LogBrewClient.create(
            apiKey = "LOGBREW_API_KEY",
            sdkName = "logbrew-kotlin",
            sdkVersion = "0.1.0",
        )
    enqueueCanonicalEvents(client)
    println(client.previewJson())
    val response =
        client.flush(
            RecordingTransport(
                listOf(
                    TransportException.network("temporary outage"),
                    202,
                ),
            ),
        )

    val helperClient = LogBrewAndroid.createClient("LOGBREW_API_KEY", "android-helper")
    val context =
        AndroidContext
            .create()
            .withActivityName("MainActivity")
            .withScreenName("Checkout")
            .withDeviceModel("Pixel")
            .withOsVersion("Android 15")
            .withSessionId("session_001")
    LogBrewAndroid.captureActivityStarted(helperClient, "evt_activity_started_001", "2026-06-02T10:00:06Z", "MainActivity", context)
    LogBrewAndroid.captureAndroidLog(
        helperClient,
        "evt_android_log_001",
        "2026-06-02T10:00:07Z",
        AndroidLogPriority.INFO,
        "Checkout",
        "button clicked",
        IllegalStateException("tap handler warning"),
        context,
    )
    LogBrewAndroid.captureThrowable(
        helperClient,
        "evt_android_exception_001",
        "2026-06-02T10:00:08Z",
        IllegalStateException("checkout failed"),
        context,
    )
    val helperPreview = helperClient.previewJson()
    check("\"activityName\": \"MainActivity\"" in helperPreview)
    check("\"androidPriority\": \"INFO\"" in helperPreview)
    check("\"throwableName\": \"IllegalStateException\"" in helperPreview)
    check("\"throwableStackTrace\"" !in helperPreview)
    check("\"source\": \"android\"" in helperPreview)

    val timelineClient = LogBrewAndroid.createClient("LOGBREW_API_KEY", "android-timeline")
    LogBrewAndroid.captureProductAction(
        timelineClient,
        "evt_android_action_001",
        "2026-06-02T10:00:09Z",
        "checkout.submit",
        context = context,
        metadata = mapOf("funnel" to "checkout", "step" to "submit", "traceId" to "trace_android_001"),
    )
    LogBrewAndroid.captureNetworkMilestone(
        timelineClient,
        "evt_android_network_001",
        "2026-06-02T10:00:10Z",
        "post",
        "/api/checkout?itemId=123#pay",
        statusCode = 503,
        durationMs = 42.5,
        context = context,
        metadata = mapOf("funnel" to "checkout", "traceId" to "trace_android_001"),
    )
    val timelinePreview = timelineClient.previewJson()
    check("\"source\": \"android.action\"" in timelinePreview)
    check("\"source\": \"android.network\"" in timelinePreview)
    check("\"name\": \"POST /api/checkout\"" in timelinePreview)
    check("\"status\": \"failure\"" in timelinePreview)
    check("\"routeTemplate\": \"/api/checkout\"" in timelinePreview)
    check("\"durationMs\": 42.5" in timelinePreview)
    check("?itemId" !in timelinePreview)
    check("#pay" !in timelinePreview)

    val metricClient = LogBrewClient.create("LOGBREW_API_KEY", "logbrew-kotlin-metrics", "0.1.0")
    metricClient.metric(
        "evt_metric_001",
        "2026-06-02T10:00:06Z",
        MetricAttributes
            .create("queue.depth", "gauge", 42.0, "{items}", "instant")
            .withMetadata(mapOf("queue" to "checkout")),
    )
    val metricPreview = metricClient.previewJson()
    check("\"type\": \"metric\"" in metricPreview)
    check("\"name\": \"queue.depth\"" in metricPreview)
    check("\"temporality\": \"instant\"" in metricPreview)
    check("\"queue\": \"checkout\"" in metricPreview)

    val httpClient =
        LogBrewClient.create(
            apiKey = "LOGBREW_API_KEY",
            sdkName = "logbrew-kotlin-http",
            sdkVersion = "0.1.0",
            maxRetries = 1,
        )
    httpClient.log(
        "evt_kotlin_http_transport",
        "2026-06-02T10:00:09Z",
        LogAttributes.create("kotlin http transport sent", "info").withLogger("kotlin-http"),
    )
    var capturedAuthorization = ""
    val httpResponse =
        httpClient.flush(
            HttpTransport(
                endpoint = "https://example.logbrew.test/v1/events",
                headers = mapOf("x-logbrew-source" to "kotlin-smoke"),
                requester =
                    HttpTransportRequester { request ->
                        capturedAuthorization = request.headers["authorization"].orEmpty()
                        check(request.headers["content-type"] == "application/json")
                        check(request.headers["x-logbrew-source"] == "kotlin-smoke")
                        check("evt_kotlin_http_transport" in request.body)
                        if (capturedAuthorization.isEmpty()) 500 else 202
                    },
            ),
        )
    check(capturedAuthorization == "Bearer LOGBREW_API_KEY")

    System.err.println(
        """{"ok":true,"status":${response.statusCode},"retryAttempts":${response.attempts},"androidHelperEvents":3,"androidTimelineEvents":2,"androidNetworkAction":"POST /api/checkout","metricEvents":1,"httpAttempts":${httpResponse.attempts}}""",
    )
}
